gremlin-go/driver/bytecode.go

```diff
@@ -19,7 +19,131 @@ under the License.
 
 package gremlingo
 
-type bytecode interface {
-	addSource(sourceName string, args ...interface{})
-	addStep(stepName string, args ...interface{})
+import "reflect"
+
+type bytecode struct {
+	sourceInstructions []instruction
+	stepInstructions   []instruction
+	bindings           map[string]interface{}
+}
+
+func newBytecode(bc *bytecode) *bytecode {
+	sourceInstructions := make([]instruction, 0)
+	stepInstructions := make([]instruction, 0)
+	bindingMap := make(map[string]interface{})
+	if bc != nil {
+		sourceInstructions = append(sourceInstructions, bc.sourceInstructions...)
+		stepInstructions = append(stepInstructions, bc.stepInstructions...)
+	}
+
+	return &bytecode{
+		sourceInstructions: sourceInstructions,
+		stepInstructions:   stepInstructions,
+		bindings:           bindingMap,
+	}
+}
+
+func (bytecode *bytecode) addSource(sourceName string, args ...interface{}) error {
```

Comment:
Since addSource and addStep are the same, would it be worth creating a function like the below and just calling it passing in the appropriate slices?
```
func (bytecode *bytecode) addStep(instructionSet []instructions, stepName string, args ...interface{}) error {
	instruction := &instruction{
		operator:  stepName,
		arguments: make([]interface{}, 0),
	}

	for _, arg := range args {
		converted, err := bytecode.convertArgument(arg)
		if err != nil {
			return err
		}
		instruction.arguments = append(instruction.arguments, converted)
	}

	instructionSet = append(instructionSet, *instruction)
	return nil
}
```